### ISSUE TITLE
@angular/localize should not depend on @angular/compiler

### ISSUE THREAD

**Issue by mhevery** [angular/angular#48163]
### Which @angular/* package(s) are the source of the bug?

localize

### Is this a regression?

No

### Description

Currently `@angular/localize` depends on `@angular/compiler` https://github.com/angular/angular/blob/main/packages/localize/src/utils/src/messages.ts#L8

`@angular/complier` is a huge package that does not execute outside of `node.js`. For this reason, it is not possible to use it with edge functions.

### Please provide a link to a minimal reproduction of the bug

https://github.com/angular/angular/blob/main/packages/localize/src/utils/src/messages.ts#L8

### Please provide the exception or error you saw

```true
`@angular/compiler` can not be loaded at edge functions such as Cloudflare, Netlify, etc...
```

### Anything else?

I think the `computeMsgId` should be in `@angular/localize` (and @angular/compiler can import it from there)

**Comment by petebacondarwin** [comment]
Although I see that it is still needed for legacy IDs... reversing the import dependency seems reasonable.

**Comment by mhevery** [comment]
I don't suppose that @petebacondarwin would be up for it?

**Comment by AndrewKushnir** [comment]
@mhevery thanks for creating this ticket!

Moving the `computeMsgId` function (and supporting code) to the `@angular/localize` and reversing the dependency (make `@angular/compiler` import `@angular/localize`) makes sense. This would also require some updates in g3, since we currently do *not* import the `@angular/localize` package. In this case, we can have a separate sub-package/entrypoint like `@angular/localize/digest` (which would only contain the message computation logic) and this would make it easier to import into g3 and minimize the code that the compiler package would depend on.

@petebacondarwin if you have some time to help with this (I see that you've assigned this ticket to yourself) - that'd be great and I can help with g3 side of things.

**Comment by petebacondarwin** [comment]
Voila! I am not sure if this counts as breaking. I thought there might be some G3 changes

**Comment by petebacondarwin** [comment]
This is the main reason we avoided doing this in the first place. There is no easy solution.

There is no real reason why the function should be owned by the compiler any more than the localize package.

Perhaps in this case we could get away with a bit of code duplication, and just have the `computeMsgId()` function copied in both packages??

**Comment by petebacondarwin** [comment]
I don't suppose you would accept a bazel package that is consumed by both the localize and compiler packages?
That way there would be duplication in the final npm packages, but not in the source?

It seems a shame that if you want to use `@angular/localize` you have to also install `@angular/compiler` just to get access to `@angular/compiler/digest` and this one function.

**Comment by mhevery** [comment]
I don't see any issue either way but:

- If we have a sub-package in the compiler for external systems such as Qwik, it is very strange that they now have to download a huge Angular compiler even though they don't need it. So aesthetically, this seems wrong. Looking forward long term if other systems would want to adopt $localize a dependency on angular compiler would be considered a problem
- My personal favorite would be to just duplicate the code.

**Comment by AndrewKushnir** [comment]
I've replied on Pete's PR (https://github.com/angular/angular/pull/48178#issuecomment-1324423545), posting here for visibility as well: it looks like code duplication is the best option for now and we'll followup with extra cleanup to the i18n subsystem afterwards.

**Comment by petebacondarwin** [comment]
Regarding removing code that was there to support ViewEngine... is there a plan to remove the "legacy ID" support from i18n? This support results in quite a lot of the `digest.ts` code that could be removed otherwise.

**Comment by petebacondarwin** [comment]
As @JoostK points out, even if we move/copy the `computeMsg()` function to `@angular/localize`, it still depends upon `@angular/compiler-cli` (due to the need for imports like `FileSystem`, `Logger`, and source-map flattening in the localize tooling.
So when you install `@angular/localize` you are still going to get both `@angular/compiler-cli` and transiently `@angular/compiler` as well.

**Comment by mhevery** [comment]
> As @JoostK points out, even if we move/copy the `computeMsg()` function to `@angular/localize`, it still depends upon `@angular/compiler-cli` (due to the need for imports like `FileSystem`, `Logger`, and source-map flattening in the localize tooling.
> So when you install `@angular/localize` you are still going to get both `@angular/compiler-cli` and transiently `@angular/compiler` as well.

Yes, all of that is true. However, the issue is not that you pull in too much code in package.json, but rather that the localized runtime pulls in `@angular/compiler` which then prevents it from running in edge functions.

So the fact that tooling requires it is not an issue. The fact that runtime requires is.

**Comment by petebacondarwin** [comment]
Does Wrangler's esbuild step not tree shape everything else??

**Comment by petebacondarwin** [comment]
In runtime mode, you are not translating the localized strings at build time but leaving the calls to `$localize` in the code.
The result is that there is code that imports `computeMsgId()` from `@angular/compiler`, that is included in the bundle (https://github.com/angular/angular/blob/c14c701775c900ce9ac8781c08fc76da067910c5/packages/localize/src/utils/src/messages.ts#L8)
So when that code that is bundled into the final application (client-side or server-side code), the bundler should only bring in that function and not the whole of the compiler, right?

What I am missing here? Is it that the compiler has side-effects that prevent any tree shaking of non-used imports?

**Comment by JoostK** [comment]
> What I am missing here? Is it that the compiler has side-effects that prevent any tree shaking of non-used imports?

Oooh, yes it does! `@angular/compiler` has `sideEffects: true` because it registers the compiler facade onto `globalThis.ng.compilerFacade`.

**Comment by petebacondarwin** [comment]
OK, so it seems that moving `computeMsgId()` to a sub-packages (e.g. `@angular/compiler/digest`) is the simplest solution. right?

**Comment by clydin** [comment]
After talking over the issue internally, we would like to avoid adding to the public API surface if possible.  Code duplication in the final output bundles of the packages shouldn't be a problem especially since the amount of code is fairly minimum (and should be even smaller soon with some upcoming refactoring).  With that in mind, a separate build target for the digest code that is then used as a build dep for each package should allow for a single repo source code location while meeting the requirement of independent package output bundles.

**Comment by DustinJSilk** [comment]
This issue seems to have gone quiet, are there any updates on this coming soon?
Thanks for the effort!

**Comment by mhevery** [comment]
Hate to be that person on the internet, but can I poke this issue to see if we could get some progress on this?

**Comment by AndrewKushnir** [comment]
Quick update: @clydin created a PR (https://github.com/angular/angular/pull/48799) which should allow the `@angular/localize` package no longer import the `@angular/compiler` package and the message digest algorithm implementation from the `@angular/compiler` package is bundled directly into the final npm package for `@angular/localize`. The PR was marked as ready for merge, so it's likely that the change will make it into one of the upcoming patch releases of Angular.

**Comment by mhevery** [comment]
Thank you!

**Comment by AndrewKushnir** [comment]
@mhevery just wanted to share a quick update: the change from PR https://github.com/angular/angular/pull/48799 was released today as a part of Angular v15.1.2. Please let us know if that helped with the original use-case. Thank you.
